### PR TITLE
Fix #50869, only cache calculated type for non-context sensitive parameters

### LINKED ISSUE #50869: Assert in assignParameterType during completionEntryDetails in reduxjs/redux-devtools
[reduxjs/redux-devtools](https://github.com/reduxjs/redux-devtools) "Parameter symbol already has a cached type [`any`] which differs from newly assigned type [an empty array/tuple, AFAICT]" while asking about the completion entry details of the rest parameter `args` [here](https://github.com/reduxjs/redux-devtools/blob/418a90914417bf3a31fc5fa4b518ae055673aca7/packages/react-base16-styling/src/index.ts#L222).

_Originally posted by @amcasey in https://github.com/microsoft/TypeScript/issues/50835#issuecomment-1252997885_

### PR DESCRIPTION
**Pull request by weswigham**
Fixes #50869

### CODE REVIEW

**Comment by weswigham** [comment]
@typescript-bot perf test this

**Comment by DanielRosenwasser** [comment]
@typescript-bot test this
@typescript-bot test top100
@typescript-bot user test this

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -9696,16 +9696,32 @@ namespace ts {
...
         function getTypeOfVariableOrParameterOrProperty(symbol: Symbol): Type {
             const links = getSymbolLinks(symbol);
             if (!links.type) {
                 const type = getTypeOfVariableOrParameterOrPropertyWorker(symbol);
                 // For a contextually typed parameter it is possible that a type has already
                 // been assigned (in assignTypeToParameterAndFixTypeParameters), and we want
-                // to preserve this type.
-                if (!links.type) {
+                // to preserve this type. In fact, we want to _prefer_ that type, and it won't
+                // be assigned until contextual typing is complete
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts:9720]:
```suggestion
                // to preserve this type. In fact, we need to _prefer_ that type, but it won't
                // be assigned until contextual typing is complete, so we need to defer in
                // cases where contextual typing may take place.
```

**Comment by DanielRosenwasser** [comment]
@typescript-bot test tsserver
@typescript-bot test tesserver top100

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -9696,16 +9696,32 @@ namespace ts {
...
+        function isParameterOfContextSensitiveSignature(symbol: Symbol) {
+            let decl = symbol.valueDeclaration;
+            if (!decl) {
+                return false;
+            }
+            if (isBindingElement(decl)) {
+                decl = walkUpBindingElementsAndPatterns(decl);
+            }
+            if (isParameter(decl)) {
+                return (isExpression(decl.parent) || isObjectLiteralMethod(decl.parent)) && isContextSensitive(decl.parent);
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts:9708]: `isExpression` is going to do a lot more work than it needs to do. We have an `isContextSensitiveFunctionOrObjectLiteralMethod` which only checks for the 3 valid forms.

There's another caveat which is that for a JSDoc'd function declaration, [we consider that to be context-sensitive-ish](https://insiders.vscode.dev/github.com/microsoft/TypeScript/blob/63791f52d4e7a3bf461b974e94abd8cbb6b546c5/src/compiler/checker.ts#L17609). I don't know if that's another scenario we have to deal with.

**Review by DanielRosenwasser — APPROVED** [review]

**Comment by DanielRosenwasser** [comment]
I wonder if this changes existing inference results - are there cases where, under `noImplicitAny`, the following will now error?

```ts
let foo = function foo(param1 = {}) {
};
```

**Comment by DanielRosenwasser** [comment]
@typescript-bot pack this

**Comment by DanielRosenwasser** [comment]
Weird that the playground build didn't work.

**Comment by DanielRosenwasser** [comment]
I guess as long as `getContextualType` does "some of the same work" you shouldn't see any divergences.

**Review by andrewbranch — APPROVED** [review]

**Comment by DanielRosenwasser** [comment]
I'd feel better if we merged this early in the 5.0 release cycle.

**Pull request merged**